components/private_channel/rust/ffi/src/lib.rs

```diff
@@ -0,0 +1,184 @@
+        pkey,
+        skey,
+        shared_pk,
+        enc_hashes,
+    } = match brave_private_channel::start_challenge(v_out, server_pk) {
+        Ok(result) => result,
+        Err(_) => return ResultChallenge::default(),
+    };
+
+    let mut pkey_buff = pkey.into_boxed_slice();
+    let pkey_ptr = pkey_buff.as_mut_ptr();
+    std::mem::forget(pkey_buff);
+
+    let mut skey_buff = skey.into_boxed_slice();
+    let skey_ptr = skey_buff.as_mut_ptr();
+    std::mem::forget(skey_buff);
+
+    let mut shared_pk_buff = shared_pk.into_boxed_slice();
+    let shared_pk_ptr = shared_pk_buff.as_mut_ptr();
+    std::mem::forget(shared_pk_buff);
+
+    let mut enc_hashes_buff = enc_hashes.into_boxed_slice();
+    let enc_hashes_size = enc_hashes_buff.clone().len();
+    let enc_hashes_ptr = enc_hashes_buff.as_mut_ptr();
+    std::mem::forget(enc_hashes_buff);
+
+    ResultChallenge {
+        pkey_ptr: pkey_ptr,
+        skey_ptr: skey_ptr,
+        key_size: KEY_SIZE,
+        shared_pubkey_ptr: shared_pk_ptr,
+        encrypted_hashes_size: enc_hashes_size,
+        encrypted_hashes_ptr: enc_hashes_ptr,
+        error: false,
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn client_second_round(
+    input: *const u8,
+    input_size: c_int,
+    client_sk_encoded: *const u8,
+) -> ResultSecondRound {
+    assert!(!input.is_null(), "Null pointers passed as input");
+    assert!(
+        !client_sk_encoded.is_null(),
+        "Null pointers passed as input"
+    );
+
+    let skey_buff = slice::from_raw_parts(client_sk_encoded, KEY_SIZE as usize);
+
+    let v_enc = slice::from_raw_parts(input, input_size as usize);
+
+    let brave_private_channel::SecondRoundOutput {
+        partial_dec,
+        proofs,
+        rand_vec,
+    } = match brave_private_channel::second_round(v_enc, skey_buff) {
+        Ok(result) => result,
+        Err(_) => return ResultSecondRound::default(),
+    };
+
+    let mut partial_enc_buff = partial_dec.into_boxed_slice();
+    let partial_enc_size = partial_enc_buff.len();
+    let partial_enc_ptr = partial_enc_buff.as_mut_ptr();
+    std::mem::forget(partial_enc_buff);
+
+    let mut proofs_buff = proofs.into_boxed_slice();
+    let proofs_size = proofs_buff.len();
+    let proofs_ptr = proofs_buff.as_mut_ptr();
+    std::mem::forget(proofs_buff);
+
+    let mut rand_vec_buff = rand_vec.into_boxed_slice();
+    let rand_vec_size = rand_vec_buff.len();
+    let rand_vec_ptr = rand_vec_buff.as_mut_ptr();
+    std::mem::forget(rand_vec_buff);
+
+    ResultSecondRound {
+        encoded_partial_dec_ptr: partial_enc_ptr,
+        encoded_partial_dec_size: partial_enc_size,
+        encoded_proofs_ptr: proofs_ptr,
+        encoded_proofs_size: proofs_size,
+        random_vec_ptr: rand_vec_ptr,
+        random_vec_size: rand_vec_size,
+        error: false,
+    }
+}
+
+#[no_mangle]
+pub unsafe extern "C" fn u8_pointer_destroy(ptr: *const u8) {
+    if !ptr.is_null() {
+        drop(ptr)
+    }
+}
```

Comment:
`drop` here is not quite like `free` in C, it will literally just forget about the pointer's existence without freeing any memory (there's no special implementation of `std::ops::Drop` for pointers, it's just treated as any value).

The best way to deallocate Rust structs is to pass them directly back and reconstruct them into the original safe Rust representations (which will be automatically `drop`ped when out of scope). In this case I'd recommend separate free methods for `ResultChallenge` and `ResultSecondRound` which take in entire instances of those structs and reconstruct the original `Vec`s using `Vec::from_raw_parts`.

The capacity argument to `Vec::from_raw_parts` just reminded me that you should also use `Vec::shrink_to_fit` on every `Vec` before passing it to C.

Reply:
- [ ] refactor destroy function to properly free the structure's content